whylog/constraints/__init__.py

```diff
@@ -2,42 +2,177 @@
+        """
+        param_dict_keys = param_dict.keys()
+        correct_param_names = self.get_param_names()
+        if not param_dict_keys == correct_param_names:
+            raise ConstructorParamsError(correct_param_names, param_dict_keys)
+        self.params = param_dict
+        self.groups = groups
+
+    @abstractmethod
+    def convert_to_user_constraint_intent(self):
+        """
+        Converts constraint to UserConstraintIntent object.
+
+        For Teacher and Config use while saving constraint into Whylog knowledge base.
+        """
+        return UserConstraintIntent(self.TYPE, self.groups, self.params)
+
+    @classmethod
+    def get_param_names(cls):
+        """
+        Returns names of constraint additional params.
+
+        For Front to display param names to user and then ask user for param contents.
+        """
+        raise NotImplementedError("Subclass should implement this")
+
+    @classmethod
+    def get_groups_count(cls):
+        """
+        Returns minimal and maximal count of groups needed to create constraint.
+
+        For Front to ask user for proper count of groups.
+        0, 0 - no groups
+        2, None - at least 2 groups
+        2, 2 - exactly 2 groups
+        """
+        return cls.MIN_GROUPS_COUNT, cls.MAX_GROUPS_COUNT
+
+    @classmethod
+    def verify(cls, param_dict, group_contents):
+        """
+        Verifies constraint for given params and groups contents.
+
+        For LogReader and Teacher verification.
+        """
+        raise NotImplementedError("Subclass should implement this")
 
 
 class TimeConstraint(AbstractConstraint):
-    def __init__(self, line_earlier, line_later, min_delta=None, max_delta=None):
-        pass
+    """
+    Time delta between two dates must be greater than 'max_delta' and lower than 'min_delta'
+    """
+
+    TYPE = ConstraintType.TIME_DELTA
+
+    MIN_GROUPS_COUNT = 0
+    MAX_GROUPS_COUNT = 0
+
+    GROUP_EARLIER = 'group_earlier'
+    GROUP_LATER = 'group_later'
+    MIN_DELTA = 'min_delta'
+    MAX_DELTA = 'max_delta'
+
+    def __init__(self, param_dict, groups):
+        """
+        I.e:
+        TimeConstraint(
+            {'group_earlier': (3, 1), 'group_later': (4, 2), 'min_delta': 12, 'max_delta' 30},
+            list()
+        )
+        """
+        super(TimeConstraint, self).__init__(param_dict, groups)
+
+    def convert_to_user_constraint_intent(self):
+        return super(TimeConstraint, self).convert_to_user_constraint_intent()
 
+    @classmethod
+    def get_param_names(cls):
+        return [cls.GROUP_EARLIER, cls.GROUP_LATER, cls.MIN_DELTA, cls.MAX_DELTA]
 
-class IdenticalIntervals(AbstractConstraint):
-    def __init__(self, intervals):
-        pass
+    @classmethod
+    def get_groups_count(cls):
+        return super(TimeConstraint, cls).get_groups_count()
 
+    @classmethod
+    def verify(cls, param_dict, group_contents):
```

Comment:
Again. Why this method takes param_dict? Why for verify just use object attributes?